src/shape.hpp

```diff
@@ -0,0 +1,117 @@
+//*****************************************************************************
+// Copyright 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <openvino/openvino.hpp>
+
+#include "shapeinfo.hpp"
+#include "status.hpp"
+
+namespace ovms {
+
+using dimension_value_t = std::int64_t;
+
+constexpr dimension_value_t DYNAMIC_DIMENSION = -1;
+
+class Dimension {
+    dimension_value_t minimum, maximum;
+
+    Dimension();
+
+public:
+    Dimension(dimension_value_t dim);
+
+    // ?????
+    Dimension(dimension_value_t minimum, dimension_value_t maximum);
+
+    bool isStatic() const;
+    bool isDynamic() const;
+
+    // TODO: Remove. For OV API 1.0 compatibility purposes only.
+    dimension_value_t getAnyValue() const;
+
+    dimension_value_t getStaticValue() const;
+    dimension_value_t getMinValue() const;
+    dimension_value_t getMaxValue() const;
+
+    bool operator==(const Dimension& rhs) const;
+    bool operator!=(const Dimension& rhs) const;
+
+    std::string toString() const;
+
+    static Dimension any();
+};
+
+class Shape {
+    std::vector<Dimension> dimensions;
+
+public:
+    Shape();
+
+    // Create shape out of ovms::Shape{1, 5, 100, 100}
+    Shape(std::initializer_list<Dimension> list);
+
+    // Create ovms::Shape out of oridnary vector of dimensions.
+    Shape(const shape_t& shape);
+
+    // Create ovms::Shape out of ov::PartialShape.
+    Shape(const ov::PartialShape& shape);
+
+    void addDimension(const Dimension& dim);
+    Shape& add(const Dimension& dim);
+
+    size_t getSize() const;
+
+    bool isStatic() const;
+    bool isDynamic() const;
+
+    ov::PartialShape createPartialShape() const;
+
+    // TODO: Remove. For OV API 1.0 compatibility purposes only.
+    shape_t getFlatShape() const;
```

Comment:
What does flat mean?

Reply:
To add dynamic shape in multiple PRs instead of one huge one we need to incrementally introduce new shape structure. To make the OVMS compile and run tests in each PR, we need old `shape_t` shape in multiple places (validation, deserialization, serialization, getmodelmetadata) for now. `get_flat_shape` converts dynamic shape to `shape_t` for this purpose. In static shape situation this is just conversion to `shape_t` bot for dynamic situation we need to _flatten_ the dimensions to remove ranges. This is simply wrong, but needs to be like that to make it compilable. This will be removed when entire dynamic shape feature is finished.